python/ql/src/semmle/python/regex.qll

```diff
@@ -124,16 +124,48 @@ abstract class RegexString extends Expr {
         )
     }
 
+    /** Escaped characters without any special handling (yet) */
+    private predicate singleEscape(int i) {
+        exists(string c |
+          c = this.getChar(i) and
+          c != "x" and c != "u" and c != "U" and c != "N"
+        )
+    }
+
+    /** Named unicode characters, eg \N{degree sign} */
+    private predicate escapedName(int start, int end) {
+        this.getChar(start + 1) = "N" and
+        this.getChar(start + 2) = "{" and
+        this.getChar(end - 1) = "}" and
+        end > start and
+        not exists(int i |
+            i > start + 2 and
+            i < end - 1 and
+            this.getChar(i) = "}"
+        )
+    }
+
     private predicate escapedCharacter(int start, int end) {
         this.escapingChar(start) and
         not exists(this.getText().substring(start + 1, end + 1).toInt()) and
         (
+            // hex value \xhh
             this.getChar(start + 1) = "x" and end = start + 4
             or
+            // octal value \ooo
             end in [start + 2 .. start + 4] and
             exists(this.getText().substring(start + 1, end).toInt())
             or
-            this.getChar(start + 1) != "x" and end = start + 2
+            // 16-bit hex value
+            this.getChar(start + 1) = "u" and end = start + 6
+            or
+            // 32-bit hex value
+            this.getChar(start + 1) = "U" and end = start + 10
+            or
+            escapedName(start, end)
+            or
+            // single character not handled above, update when adding a new case
+            this.singleEscape(start + 1) and end = start + 2
```

Comment:
I wonder if this does the right thing in all cases. For instance, `\s` is not really an escaped character, but a representative of an entire character class. I can't immediately see that this will cause a problem, but it's something to consider. Likewise, `\A`, `\b`, `\B`, and `\Z` all match zero characters, so it might be sensible to exclude them from this predicate.

Reply:
Excluding these from the predicate makes sense to me, but changes the expected files in a way I might need some help evaluating.

Reply:
I created [an issue](https://github.com/github/codeql-python-team/issues/100) to track this and propose to solve it at a later stage.